rust/lance-index/src/vector/ivf.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! IVF - Inverted File Index
+
+use std::ops::Range;
+use std::sync::Arc;
+
+use arrow_array::{
+    cast::AsArray,
+    types::{Float32Type, UInt32Type},
+    Array, ArrayRef, Float32Array, RecordBatch, UInt32Array,
+};
+use arrow_ord::sort::sort_to_indices;
+use arrow_schema::Field;
+use arrow_select::take::take;
+use lance_arrow::RecordBatchExt;
+use lance_core::{Error, Result};
+use lance_linalg::{
+    distance::{Cosine, Dot, MetricType, Normalize, L2},
+    kernels::argmin,
+    MatrixView,
+};
+use snafu::{location, Location};
+use tracing::instrument;
+
+use super::PART_ID_COLUMN;
+use crate::vector::transform::Transformer;
+
+/// IVF - IVF file partition
+///
+#[derive(Debug, Clone)]
+pub struct Ivf {
```

Comment:
This is duplicating the `Ivf` struct that's in the `lance` crate right? Or was this supposed to be a move?

Reply:
I am going to move the IVF model from lance to here. Wanted to get this in before it becomes too big.